Add AK/IPAddressCidr datatype

famfo opened the pull request:
TODO:
 - IPv6AddressCidr
 - IPv6AddressCidr tests

famfo committed:
WIP: AK: Introduce IPAddressCidr datatype

This commit introduces the IPAddressCidr datatype, it is meant to
provide a more up to date alternative to the netmask notation.
IPv4AddressCidr and tests for it are complete, IPv6AddressCidr is
still TODO.

kleinesfilmroellchen commented on AK/IpAddressCidr.h:
```diff
@@ -0,0 +1,175 @@
+        StringParsingFailed,
+    };
+
+    static Result<AddressFamily, IPAddressCidrError> Create(AddressTraits<AddressFamily>::IpAddress address, u8 length)
```
Result is kinda deprecated (not sure why not removed yet), we use ErrorOr.

kleinesfilmroellchen replied:
Also, no capital function names. `create` is fine.

kleinesfilmroellchen commented on AK/IpAddressCidr.h:
```diff
@@ -0,0 +1,175 @@
+template<OneOf<IPv4AddressCidr, IPv6AddressCidr> AddressFamily>
+class [[gnu::packed]] IPAddressCidr {
+public:
+    enum IPAddressCidrError {
```
always enum class

kleinesfilmroellchen commented on AK/IpAddressCidr.h:
```diff
@@ -0,0 +1,175 @@
+    using IpAddress = IPv4Address;
+};
+
+class [[gnu::packed]] IPv4AddressCidr : public IPAddressCidr<IPv4AddressCidr> {
```
almost all functions in here can be constexpr

kleinesfilmroellchen commented on AK/IpAddressCidr.h:
```diff
@@ -0,0 +1,175 @@
+        return AddressFamily(address, length);
+    }
+
+    AddressTraits<AddressFamily>::IpAddress ip_address() const { return m_address; }
```
I’m fairly sure you want to alias the dependent IP address type with `using IPAddress = AddressTraits<AddressFamily>::IPAddress` at the start of the public section of this base class.

kleinesfilmroellchen commented on AK/IpAddressCidr.h:
```diff
@@ -0,0 +1,175 @@
+
+namespace AK {
+
+class [[gnu::packed]] IPv4AddressCidr;
+class [[gnu::packed]] IPv6AddressCidr;
+template<typename Address>
+class [[gnu::packed]] AddressTraits;
+
+template<OneOf<IPv4AddressCidr, IPv6AddressCidr> AddressFamily>
+class [[gnu::packed]] IPAddressCidr {
```
This can go into a Details {} namespace since I think noone wants to access it directly.

kleinesfilmroellchen replied:
In all cases, using `[[gnu::packed]]` is unnecessary here, since this type is not transmitted over the network and we don’t need particular small alignment guarantees. This is only done for the other IP-related types that actually go out to hardware.

kleinesfilmroellchen commented on AK/IpAddressCidr.h:
```diff
@@ -0,0 +1,175 @@
+        return first_address_of_subnet() == other_cidr.first_address_of_subnet();
+    }
+
+    static u8 const MAX_LENGTH = 32;
```
this is `constexpr` (which will allow you to make a lot of other stuff constexpr)

kleinesfilmroellchen reviewed (approved):
please squash commits :)

famfo commented:
squashed the commits

famfo committed:
AK/IpAddressCidr: Add IPv6AddressCidr

famfo committed:
AK/IpAddressCidr: Fix Endianness probles, add testcases

famfo committed:
AK/IpAddressCidr: Fix Kernel define

This commit:
- includes Kernel/Library/MiniStdLib.h for the memcpy function
- changes the to_string() function to append the KString->view() to
  the string builder which is a type that can be appended by
  try_append
